js/cpu6502.ts

```diff
@@ -0,0 +1,1759 @@
+/*
+ * Copyright 2010-2019 Will Scullin <scullin@scullinsteel.com>
+ *
+ * Permission to use, copy, modify, distribute, and sell this software and its
+ * documentation for any purpose is hereby granted without fee, provided that
+ * the above copyright notice appear in all copies and that both that
+ * copyright notice and this permission notice appear in supporting
+ * documentation.  No representations are made about the suitability of this
+ * software for any purpose.  It is provided "as is" without express or
+ * implied warranty.
+ */
+
+import { byte, word } from './types';
+import { debug, toHex } from './util';
+
+type symbols = { [key: number]: string };
+
+export interface CpuOptions {
+    '65C02'?: boolean;
+}
+
+export interface CpuState {
+        a: byte,
+        x: byte,
+        y: byte,
+        s: byte,
+        pc: word,
+        sp: byte,
+        cycles: number
+}
+
+/** Range of mode numbers. */
+type mode = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15;
+
+/** Addressing mode name to number mapping. */
+const modes: { [key: string]: mode } = {
+    accumulator: 0,           // A (Accumulator)
+    implied:     1,           // Implied
+    immediate:   2,           // # Immediate
+    absolute:    3,           // a Absolute
+    zeroPage:    4,           // zp Zero Page
+    relative:    5,           // r Relative
+
+    absoluteX:   6,           // a,X Absolute, X
+    absoluteY:   7,           // a,Y Absolute, Y
+    zeroPageX:   8,           // zp,X Zero Page, X
+    zeroPageY:   9,           // zp,Y Zero Page, Y
+
+    absoluteIndirect:   10,  // (a) Indirect
+    zeroPageXIndirect:  11, // (zp,X) Zero Page Indexed Indirect
+    zeroPageIndirectY:  12, // (zp),Y Zero Page Indexed with Y
+
+    /* 65c02 */
+    zeroPageIndirect:   13,  // (zp),
+    absoluteXIndirect:  14,  // (a, X),
+    zeroPage_relative:  15   // zp, Relative
+};
+
+/** Instruction size by addressing mode. */
+const sizes = {
+    0 /* modes.accumulator */: 1,
+    1 /* modes.implied */: 1,
+    2 /* modes.immediate */: 2,
+    3 /* modes.absolute */: 3,
+    4 /* modes.zeroPage */: 2,
+    5 /* modes.relative */: 2,
+    6 /* modes.absoluteX */: 3,
+    7 /* modes.absoluteY */: 3,
+    8 /* modes.zeroPageX */: 2,
+    9 /* modes.zeroPageY */: 2,
+    10 /* modes.indirect */: 3,
+    11 /* modes.zeroPageXIndirect */: 2,
+    12 /* modes.zeroPageYIndirect */: 2,
+
+    13 /* mode.zeroPageIndirect */: 2,
+    14 /* mode.absoluteXIndirect */: 3,
+    15 /* mode.zeroPage_relative */: 3
+};
```

Comment:
I was considering moving the disassembler to its own class. It would lead to some duplication, though. Thoughts?

Reply:
It would probably be minimal duplication, The parts of the instruction table that the cpu uses and the disassembler uses don't really overlap anymore since I attempted cycle correctness. It's possibly a little handy to have the instruction there sometimes, but probably not necessary.